recipients_app/lib/data/repositories/user_repository.dart

```diff
@@ -1,94 +1,79 @@
+import "package:app/data/datasource/remote/user_remote_data_source.dart";
+import "package:app/data/datasource/user_data_source.dart";
 import "package:app/data/models/models.dart";
-import "package:app/data/repositories/repositories.dart";
+import "package:app/demo_manager.dart";
 import "package:cloud_firestore/cloud_firestore.dart";
 import "package:firebase_auth/firebase_auth.dart";
 
-const String recipientCollection = "/recipients";
-
 class UserRepository {
+  late UserDataSource remoteDataSource = UserRemoteDataSource(firestore: firestore, firebaseAuth: firebaseAuth);
+  late UserDataSource demoDataSource = UserDemoDataSource();
+
+  final DemoManager demoManager;
   final FirebaseFirestore firestore;
   final FirebaseAuth firebaseAuth;
 
-  const UserRepository({
+  UserRepository({
     required this.firestore,
     required this.firebaseAuth,
+    required this.demoManager,
   });
 
-  Stream<User?> authStateChanges() => firebaseAuth.authStateChanges();
-  User? get currentUser => firebaseAuth.currentUser;
-
-  /// Fetches the user data by userId from firestore and maps it to a recipient object
-  /// Returns null if the user does not exist.
-  Future<Recipient?> fetchRecipient(User firebaseUser) async {
-    final phoneNumber = firebaseUser.phoneNumber ?? "";
-
-    final matchingUsers = await firestore
-        .collection(recipientCollection)
-        .where(
-          "mobile_money_phone.phone",
-          isEqualTo: int.parse(phoneNumber.substring(1)),
-        )
-        .get();
-
-    if (matchingUsers.docs.isEmpty) {
-      return null;
-    }
-
-    final userSnapshot = matchingUsers.docs.firstOrNull;
-
-    // This doesnt work because user id from firebaseAuth is not related to user id from firestore
-    // Needs to be discussed if changes should be made or not
-    // final userSnapshot =
-    //     await firestore.collection("/recipients").doc(firebaseUser.uid).get();
-
-    if (userSnapshot != null && userSnapshot.exists) {
-      // TODO: decide if we should keep it in user object in the app at all
-      final payments =
-          await PaymentRepository(firestore: firestore).fetchPayments(recipientId: userSnapshot.id);
-
-      return Recipient.fromMap(userSnapshot.data()).copyWith(
-        payments: payments,
-        userId: userSnapshot.id,
-      );
-    } else {
-      return null;
-    }
+  UserDataSource get _activeDataSource => demoManager.isDemoEnabled ? demoDataSource : remoteDataSource;
+
+  Stream<User?> authStateChanges() {
+    final StreamController<User?> authStateController = StreamController();
+    StreamSubscription<User?>? authStateSubscription;
+
+    authStateSubscription = _activeDataSource.authStateChanges().listen((authState) {
+      authStateController.add(authState);
+    });
+
+    demoManager.isDemoEnabledStream.listen((isDemoMode) {
+      authStateSubscription?.cancel();
+
+      authStateSubscription = _activeDataSource.authStateChanges().listen((authState) {
```

Comment:
I'm not sure if this logic is right.
You listen here for changes on the demo mode flag. If demo mode is true, then you start listen for auth changes on the activeDataSource which is the "UserDemoDataSource". So far, so good.

But you do the same code if demo mode is false. And is this right? Or should be only do "authStateSubscription?.cancel();" in this case??????

Reply:
Ah, I got it :) It's right. This code ensures that we update the authStateController if the active data source is changing. Right?